mlir/include/mlir/Interfaces/MemorySlotInterfaces.h

```diff
@@ -23,6 +23,33 @@ struct MemorySlot {
   Type elemType;
 };
 
+struct DestructibleSlotInfo {
+  /// Maps from an index within the memory slot the type of the pointer that
+  /// will be generated to access the element directly.
+  DenseMap<Attribute, Type> elementsPtrs;
+};
+
+struct SubElementSlotInfo {
+  /// Index of this memory slot in the parent memory slot.
+  Attribute subelementIndex;
+};
+
+struct DestructibleMemorySlot {
+  MemorySlot slot;
+  DestructibleSlotInfo info;
+};
+
+struct SubElementMemorySlot {
```

Comment:
This has been removed now, but I did not go with inheritance because that led to uncomfortable diamond-shaped inheritance graphs. Composition works best in that context.